src/model.cpp

```diff
@@ -189,4 +214,74 @@ void model::set_local_spike_callback(spike_export_function export_callback) {
     local_export_callback_ = export_callback;
 }
 
+util::optional<cell_size_type> model::local_cell_index(cell_gid_type gid) {
+    auto it = gid_to_local_.find(gid);
+    return it==gid_to_local_.end()?
+        util::nothing:
+        util::optional<cell_size_type>(it->second);
+}
+
+void model::inject_events(const event_vector& events) {
+    using pse = postsynaptic_spike_event;
+
+    auto& lanes = event_lanes(epoch_.id);
+
+    // Append all events that are to be delivered to local cells to the
+    // appropriate lane. At the same time, keep track of which lanes have been
+    // modified, because the lanes will have to be sorted once all events have
+    // been added.
+    event_vector local_events;
+    std::set<cell_size_type> modified_lanes;
+    for (auto& e: events) {
+        if (e.time<t_) {
+            throw std::runtime_error("model::inject_events(): attempt to inject an event at time " + std::to_string(e.time) + ", when model state is at time " + std::to_string(t_));
+        }
+        if (auto lidx = local_cell_index(e.target.gid)) {
+            lanes[*lidx].push_back(e);
+            modified_lanes.insert(*lidx);
+        }
+    }
+
+    // Sort events in the event lanes that were modified
+    auto event_bind = [] (const pse& e) {return std::tie(e.time, e.target, e.weight);};
+    for (auto l: modified_lanes) {
+        util::sort_by(lanes[l], event_bind);
+    }
+}
+
+// Merge events that are to be delivered from two lists into a sorted list.
+// Events are sorted by delivery time, then target, then weight.
+//
+//  tfinal: The time at which the current epoch finishes.
+//  lc: Sorted set of events to be delivered before and after `tfinal`.
+//  events: Unsorted list of events with delivery time greater than or equal to
+//      tfinal. May be modified by the call.
+//  lf: Will hold a list of all postsynaptic events in `events` and `lc` that
+//      have delivery times greater than or equal to `tfinal`.
+void merge_events(time_type tfinal, const event_vector& lc, event_vector& events, event_vector& lf) {
+    using pse = postsynaptic_spike_event;
+
+    // For ordering events by: delivery time, then target index, then weight.
+    auto lex_bind = [](const pse& e){return std::tie(e.time, e.target, e.weight);};
+    auto lex_less = [lex_bind] (const pse& l, const pse& r) {return lex_bind(l)<lex_bind(r);};
+
+    // Merge the incoming events with events in lc that are not to be delivered
+    // in this epoch, and store the result in lf.
+
+    // STEP 1: sort events in place in events[l]
+    PE("sort");
+    util::sort_by(events, lex_bind);
```

Comment:
We've got `lex_bind` here, but `event_bind` in `model::inject_events`. The bind is only used for sorting as far as I can tell. Can we define instead either a `pse_less` that does this comparison, or make it the `operator<` overload for `postsynaptic_spike_event`?

Reply:
I changed the name of `lex_bind` and `lex_less` to `pse_tie` and `pse_less`, and  described the lambda `event_bind` when passing it in `inject_events`.
I still need both the `tie` and the `less` operators for `sort_by` and `merge` respectively, so I would rather define both in line, where needed here.

Reply:
If we haven't already got a range-like sort wrapper, it'd be easy to add one — this could obviate the need for both a bind and a less.

Reply:
I will add one.